storage/remote/queue_manager.go

```diff
@@ -166,79 +177,92 @@ func (t *StorageQueueManager) Describe(ch chan<- *prometheus.Desc) {
+		queueLength += len(shard)
+	}
+	return queueLength
+}
+
 // Collect implements prometheus.Collector.
 func (t *StorageQueueManager) Collect(ch chan<- prometheus.Metric) {
 	t.samplesCount.Collect(ch)
 	t.sendLatency.Collect(ch)
-	t.queueLength.Set(float64(len(t.queue)))
+	t.queueLength.Set(float64(t.QueueLength()))
 	ch <- t.failedBatches
 	ch <- t.failedSamples
 	ch <- t.queueLength
 	ch <- t.queueCapacity
 }
 
-func (t *StorageQueueManager) sendSamples(s model.Samples) {
-	t.sendSemaphore <- true
-
-	go func() {
-		defer func() {
-			<-t.sendSemaphore
-		}()
-
-		// Samples are sent to the remote storage on a best-effort basis. If a
-		// sample isn't sent correctly the first time, it's simply dropped on the
-		// floor.
-		begin := time.Now()
-		err := t.tsdb.Store(s)
-		duration := time.Since(begin).Seconds()
-
-		labelValue := success
-		if err != nil {
-			log.Warnf("error sending %d samples to remote storage: %s", len(s), err)
-			labelValue = failure
-			t.failedBatches.Inc()
-			t.failedSamples.Add(float64(len(s)))
-		}
-		t.samplesCount.WithLabelValues(labelValue).Add(float64(len(s)))
-		t.sendLatency.Observe(duration)
-	}()
-}
-
 // Run continuously sends samples to the remote storage.
 func (t *StorageQueueManager) Run() {
-	defer func() {
-		close(t.drained)
-	}()
+	for i := 0; i < t.cfg.Shards; i++ {
+		go t.runShard(i)
+	}
+	t.wg.Wait()
+}
 
-	// Send batches of at most maxSamplesPerSend samples to the remote storage.
+// Stop stops sending samples to the remote storage and waits for pending
+// sends to complete.
+func (t *StorageQueueManager) Stop() {
+	log.Infof("Stopping remote storage...")
+	for _, shard := range t.shards {
+		close(shard)
+	}
+	t.wg.Wait()
+	log.Info("Remote storage stopped.")
+}
+
+func (t *StorageQueueManager) runShard(i int) {
+	defer t.wg.Done()
+	shard := t.shards[i]
+
+	// Send batches of at most MaxSamplesPerSend samples to the remote storage.
 	// If we have fewer samples than that, flush them out after a deadline
 	// anyways.
+	pendingSamples := model.Samples{}
+
 	for {
 		select {
-		case s, ok := <-t.queue:
+		case s, ok := <-shard:
 			if !ok {
-				log.Infof("Flushing %d samples to remote storage...", len(t.pendingSamples))
-				t.flush()
+				log.Infof("Flushing %d samples to remote storage...", len(pendingSamples))
+				t.sendSamples(pendingSamples)
 				log.Infof("Done flushing.")
 				return
 			}
 
-			t.pendingSamples = append(t.pendingSamples, s)
+			pendingSamples = append(pendingSamples, s)
 
-			for len(t.pendingSamples) >= maxSamplesPerSend {
-				t.sendSamples(t.pendingSamples[:maxSamplesPerSend])
-				t.pendingSamples = t.pendingSamples[maxSamplesPerSend:]
+			for len(pendingSamples) >= t.cfg.MaxSamplesPerSend {
+				t.sendSamples(pendingSamples[:t.cfg.MaxSamplesPerSend])
+				pendingSamples = pendingSamples[t.cfg.MaxSamplesPerSend:]
 			}
-		case <-time.After(batchSendDeadline):
-			t.flush()
+		case <-time.After(t.cfg.BatchSendDeadline):
+			t.sendSamples(pendingSamples)
```

Comment:
You deleted `flush()`, but I think you'll still want the check it contained for an empty list of pending samples. Otherwise we're going to send empty batches on every deadline.